app/bundles/AssetBundle/EventListener/BuilderSubscriber.php

```diff
@@ -119,30 +124,23 @@ public function onPageDisplay(PageDisplayEvent $event)
     }
 
     /**
-     * @param       $event
-     * @param       $leadId
-     * @param array $source
-     * @param null  $emailId
+     * @param PageDisplayEvent|EmailSendEvent $event
+     * @param int                             $leadId
+     * @param array                           $source
+     * @param null                            $emailId
      *
      * @return array
      */
     private function generateTokensFromContent($event, $leadId, $source = [], $emailId = null)
     {
-        $content = $event->getContent();
-
-        $clickthrough = [];
         if ($event instanceof PageDisplayEvent || ($event instanceof EmailSendEvent && $event->shouldAppendClickthrough())) {
-            $clickthrough = ['source' => $source];
-
-            if ($leadId !== null) {
-                $clickthrough['lead'] = $leadId;
-            }
-
-            if (!empty($emailId)) {
-                $clickthrough['email'] = $emailId;
-            }
+            $clickthrough = [
+                'source' => $source,
+                'lead'   => $leadId ?? false,
+                'email'  => $emailId ?? false,
+            ];
         }
 
-        return $this->tokenHelper->findAssetTokens($content, $clickthrough);
+        return $this->tokenHelper->findAssetTokens($event->getContent(), array_filter($clickthrough ?? []));
```

Comment:
This was a change that @dongilbert made. My understanding of it is that without the callback, array_filter checks the true/false value of each item in the array, so it will only return array items with truthy value.

Reply:
I've checked in the [documentation](https://www.php.net/manual/en/function.array-filter.php) as well. It reads:

> If no callback is supplied, all entries of array equal to FALSE (see converting to boolean) will be removed.

The `will be removed` part concerns me. Do I understand it right that is deprecated? @dongilbert do you have other explanation of what they mean by that?

Reply:
The previous code would only set values on the clickthrough array if they weren't empty (or equal to false, given the check for `!empty()`. We can save on that 8 lines of code checking if they are empty by setting them regardless of their value, then filtering the array with array_filter. Without supplying the callback, empty values in the array will be filtered out. I don't understand your deprecated comment @escopecz

Reply:
Ah, I get it. No - that functionality is not deprecated. It's staging that array values equal to false will be removed, not that the functionality will be removed.

Reply:
Ok, I get it now. Thanks!